### PR TITLE
Teuchos/Tpetra: Fix for #9572

### LINKED ISSUE #9572: Ifpack2:  Build error mixing complex<float> and complex<double> in Thyra_Ifpack2PreconditionerFactory_def.hpp
## Bug Report
@trilinos/Ifpack2 @trilinos/Thyra

### Description
What went wrong?  What should have happened?  Do you have an idea what might
fix things?  How has this bug affected you?  What are you trying to accomplish?
What specifically must we demonstrate to call this work complete?

### Steps to Reproduce
1. SHA1:  db2b7fb220f3853fd7bd325775a3c51f908b54f3 (current develop)

Build error is:
```
/scratch/vebruni/src/code/objs/tpls/spack/spack/.spack/stage/spack-stage-trilinos-2021.08.16-qi2tfvfw2j3jdwl6pqwi5jfail4q2m6y/spack-src/packages/tpetra/core/src/Tpetra_MixedScalarMultiplyOp.hpp:125:35: error: no viable conversion from 'complex<double>' to 'complex<float>'
      op_->apply (*X_, *Y_, mode, alpha, beta);
                                  ^~~~~
/scratch/vebruni/src/code/objs/tpls/spack/spack/.spack/stage/spack-stage-trilinos-2021.08.16-qi2tfvfw2j3jdwl6pqwi5jfail4q2m6y/spack-src/packages/tpetra/core/src/Tpetra_MixedScalarMultiplyOp.hpp:94:5: note: in instantiation of member function 'Tpetra::MixedScalarMultiplyOp<std::complex<double>, std::complex<float>, int, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::apply' requested here
    MixedScalarMultiplyOp (const Teuchos::RCP<const op_type>& A) :
    ^
/scratch/vebruni/src/code/objs/tpls/spack/spack/.spack/stage/spack-stage-trilinos-2021.08.16-qi2tfvfw2j3jdwl6pqwi5jfail4q2m6y/spack-src/packages/ifpack2/adapters/thyra/Thyra_Ifpack2PreconditionerFactory_def.hpp:252:42: note: in instantiation of member function 'Tpetra::MixedScalarMultiplyOp<std::complex<double>, std::complex<float>, int, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::MixedScalarMultiplyOp' requested here
    RCP<TpetraLinOp> wrappedOp = rcp(new Tpetra::MixedScalarMultiplyOp<scalar_type,half_scalar_type,local_ordinal_type,global_ordinal_type,node_type>(concretePrecOpHalf));
                                         ^
/projects/sierra/linux_rh7/SDK/compilers/gcc/7.2.0-RHEL7/lib/gcc/x86_64-pc-linux-gnu/7.2.0/../../../../include/c++/7.2.0/complex:1051:12: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'std::complex<double>' to 'const std::complex<float> &' for 1st argument
    struct complex<float>
           ^
/projects/sierra/linux_rh7/SDK/compilers/gcc/7.2.0-RHEL7/lib/gcc/x86_64-pc-linux-gnu/7.2.0/../../../../include/c++/7.2.0/complex:1051:12: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'std::complex<double>' to 'std::complex<float> &&' for 1st argument
/projects/sierra/linux_rh7/SDK/compilers/gcc/7.2.0-RHEL7/lib/gcc/x86_64-pc-linux-gnu/7.2.0/../../../../include/c++/7.2.0/complex:1056:26: note: candidate constructor not viable: no known conversion from 'std::complex<double>' to 'std::complex<float>::_ComplexT' (aka '_Complex float') for 1st argument
      _GLIBCXX_CONSTEXPR complex(_ComplexT __z) : _M_value(__z) { }
                         ^
/projects/sierra/linux_rh7/SDK/compilers/gcc/7.2.0-RHEL7/lib/gcc/x86_64-pc-linux-gnu/7.2.0/../../../../include/c++/7.2.0/complex:1058:26: note: candidate constructor not viable: no known conversion from 'std::complex<double>' to 'float' for 1st argument
      _GLIBCXX_CONSTEXPR complex(float __r = 0.0f, float __i = 0.0f)
                         ^
/projects/sierra/linux_rh7/SDK/compilers/gcc/7.2.0-RHEL7/lib/gcc/x86_64-pc-linux-gnu/7.2.0/../../../../include/c++/7.2.0/complex:1068:35: note: explicit constructor is not a candidate
      explicit _GLIBCXX_CONSTEXPR complex(const complex<double>&);
                                  ^
/projects/sierra/linux_rh7/SDK/compilers/gcc/7.2.0-RHEL7/lib/gcc/x86_64-pc-linux-gnu/7.2.0/../../../../include/c++/7.2.0/complex:1069:35: note: explicit constructor is not a candidate
      explicit _GLIBCXX_CONSTEXPR complex(const complex<long double>&);
                                  ^
/scratch/vebruni/src/code/objs/tpls/spack/spack/.spack/stage/spack-stage-trilinos-2021.08.16-qi2tfvfw2j3jdwl6pqwi5jfail4q2m6y/spack-src/packages/tpetra/core/src/Tpetra_Operator.hpp:123:19: note: passing argument to parameter 'alpha' here
           Scalar alpha = Teuchos::ScalarTraits<Scalar>::one(),
```

### PR DESCRIPTION
**Pull request by cgcgcg**
@trilinos/teuchos @trilinos/tpetra 

## Motivation
Fix #9572

### CODE REVIEW

**Review by bartlettroscoe — COMMENTED** [review]
Technically this should have a unit test in Teuchos as this is Teuchos code.  We want to move to a Trilinos that has more independently sustainable packages and that means that packages need to be testing their own code.

**Review by jhux2 — APPROVED** [review]
lgtm

**Comment by cgcgcg** [comment]
@bartlettroscoe I added a test for the new type conversion. May I merge?

File: packages/teuchos/core/src/Teuchos_as.hpp [hunk 1]
```diff
@@ -2576,6 +2576,21 @@ class ValueTypeConversionTraits<std::complex<RealType>, unsigned long long> {
...
 };
 
+
+//! Convert std::complex<double> to std::complex<float>.
+template<>
+class ValueTypeConversionTraits<std::complex<float>, std::complex<double> > {
+public:
+  inline static std::complex<float> convert (const std::complex<double> t) {
+    // Let double->float conversion handle real and imaginary part.
+    return std::complex<float> (as<float> (t.real ()), as<float> (t.imag ()));
```
Comment (bartlettroscoe) [inline packages/teuchos/core/src/Teuchos_as.hpp:2586]: I suspect that if you change this code to put in `0.0` for the `imag` part, that the Teuchos unit tests will still pass.

File: packages/teuchos/core/src/Teuchos_as.hpp [hunk 2]
```diff
@@ -2576,6 +2576,21 @@ class ValueTypeConversionTraits<std::complex<RealType>, unsigned long long> {
...
+  }
+  static std::complex<float> safeConvert (const std::complex<double> t) {
+    // Let double->float conversion handle real and imaginary part.
+    return std::complex<float> (asSafe<float> (t.real ()), asSafe<float> (t.imag ()));
```
Comment (bartlettroscoe) [inline packages/teuchos/core/src/Teuchos_as.hpp:2590]: Same here.  I think if you were to put in `0.0.` for the `imag` part, the Teuchos unit tests would still pass.

File: packages/teuchos/core/test/TypeConversions/TypeConversions_UnitTest.cpp [hunk 3]
```diff
@@ -1057,6 +1057,78 @@ UNIT_TEST_GROUP_SIGNED_INTEGER( long_long_type )
...
 
+
+#ifdef HAVE_TEUCHOS_COMPLEX
+
+#include <complex>
+
+//
+// Tests for conversions involving complex types.
+//
+TEUCHOS_UNIT_TEST( asSafe, complexToComplex ) {
```
Comment (bartlettroscoe) [inline packages/teuchos/core/test/TypeConversions/TypeConversions_UnitTest.cpp:1068]: I don't think these tests ever test the imaginary part, do they?

**Review by bartlettroscoe — COMMENTED** [review]
@cgcgcg, thanks for adding the Teuchos unit test.  We can see that test running and passing the the last PR testing iteration in this query:

* https://testing.sandia.gov/cdash/queryTests.php?project=Trilinos&filtercount=4&showfilters=1&filtercombine=and&field1=buildname&compare1=65&value1=PR-9575-test-Trilinos_&field2=testname&compare2=61&value2=TeuchosCore_TypeConversions_UnitTest_MPI_1&field3=testoutput&compare3=95&value3=asSafe_complexToComplex_UnitTest&field4=buildstarttime&compare4=84&value4=now

showing:

image

For some reason, that test only runs in 4 builds but that is a good set of builds with GCC debug on and off and CUDA so that is good enough.

However, I think these tests are only testing the real part of the conversion for the complex numbers.  Don't you need at least one check that sets the imaginary part?  I think if you create the vars:

```
const std::complex<float> minusOnePlusTwoF(-1,+2);
const std::complex<double> minusOnePlusTwoD(-1, +2);
```

and did conversions with those as well, then that at least test that the imaginary part was being copied correctly.

**Comment by cgcgcg** [comment]
@bartlettroscoe You're right, that is better. I added some imaginary checks. It's surprising that we have more than one build that enables complex :-)

**Review by bartlettroscoe — APPROVED** [review]
@cgcgcg, thanks for updating  unit tests for checking imaginary part.

**Pull request merged**